src/ir/auth_logic/type_environment.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/ir/auth_logic/type_environment.h"
+
+#include "src/ir/auth_logic/is_name_constant.h"
+
+namespace raksha::ir::auth_logic {
+
+namespace {
+
+// This visitor aids in the construction of the TypeEnvironment.
+// It populates a mapping (flat_hash_map) of constant (literal)
+// names to their types by looking at the relations to which the literals
+// are applied and declarations of those relations (including the types
+// of the parameters where they are applied). It takes as input
+// a DeclarationEnvironment.
+// This is a nested class because it needs to access the private
+// fields of the TypeEnvironment in the constructor implementation.
+
+using LiteralTypeMapType =
```

Comment:
Not a huge problem as this is in an anonymous namespace, but would be good to move it inside the class?

Reply:
fixed